0018-logging.adoc

```diff
@@ -0,0 +1,94 @@
+= [Overhaul comit-rs Logging]
+Tobin C. Harding <tobin.harding@coblox.tech>;
+:toc:
+:revdate: 2019-12-23
+
+NOTE: Author: {authors} +
+Date: {revdate} +
+Tracking issue: https://github.com/comit-network/comit-rs/issues/487[#487]
+
+== Context
+
+At the moment, there is no consistency in what is logged where and with what level of detail.
+Debugging cnd by using the logs is painful at best and impractical at worst.
+
+This spike is about investigating, how the current logs could be improved and proposing an overall logging strategy that makes the logs actually useful.
+
+There are a few problems that need addressing, including but not limited to:
+
+* Log levels - guidelines on when to log and at what level
+* Useful context - we need useful context, what is useful should be defined by the team since it comes from experience debugging cnd.
+* Only useful output i.e., no dumping u8 arrays in the logs.
+* Async codebase - since we have an asynchronous code base logging may be inherently the wrong approach, we will also consider tracing.
+* Dynamically controlling log level
+* Controlling the log level of libraries vs cnd logs i.e., 
+
+The output of this spike should be proposed changes to the current logging libraries (i.e., what libraries we use to log) and also some guidelines so the whole team knows when and where to add log statements.
+
+== Research
+
+=== Log Levels ===
+
+
+This is actually the easiest thing to solve, using correct log levels is pretty standard dev practice, the internet more or less has consensus on the following log levels:
+
+Copying from the following site: https://reflectoring.io/logging-levels/
+
+* error: should only be used when the application really is in trouble. Users are being affected without having a way to work around the issue.
+* warn: should be used when something bad happened, but the application still has the chance to heal itself or the issue can wait a day or two to be fixed.
+* info: should be used to document state changes in the application or some entity within the application.
+* debug: should be used to log any information that helps us identify what went wrong on DEBUG level.
+* trace: log all information that helps us to trace the flow of the application.
+
+=== Useful Context ===
+
+Useful context is not that easy to solve due to the async nature of cnd.
```

Comment:
I would expect the swap id being mentioned here. This is the one information I would like to see on all log lines.

Reply:
If you want the swap_id on every log line then I suggest we just drop this whole discussion and go for tracing (where we get context for free) - adding this to every log line will probably lead to
- a macro that wraps the log statements
- passing extra parameters around so we always have access to the swap id when we want to log.

Reply:
I expect the swap_id to be present at every line yes. Otherwise it means our logging because useless as soon as someone runs 2 swap at the same time.